cadccutout/cadccutout/file_helpers/fits/fits_file_helper.py

```diff
@@ -162,43 +162,53 @@ def _pixel_cutout(self, hdu, cutout_dimension):
             data=hdu.data, cutout_dimension=cutout_dimension, wcs=wcs)
 
         self._post_sanitize_header(header, cutout_result)
-        self._write_cutout(header=header, data=cutout_result.data)
-
-    def _write_primary_hdu(self, hdu_list):
-        primary_hdu = hdu_list[0]
-        if isinstance(primary_hdu, PrimaryHDU):
-            logger.debug('Writing out Primary HDU.')
-            fits.append(
-                filename=self.output_writer, header=primary_hdu.header,
-                data=None, overwrite=False, output_verify='silentfix',
-                checksum='remove')
-            logger.debug('Wrote out Primary HDU.')
+        result = fits.ImageHDU(header=header, data=cutout_result.data)
+        self._fix_header(result.header, header)
+        return result
+
+    def _fix_header(self, target, original):
+        # overcome astropy's tendency to remove keywords from original headers
+        if 'BSCALE' in original:
+            target['BSCALE'] = original['BSCALE']
+        if 'BZERO' in original:
+            target['BZERO'] = original['BZERO']
+
+    def _add_primary_hdu(self, hdu_list, result_hdu_list):
+        if isinstance(hdu_list[0], PrimaryHDU):
+            logger.debug('Appending Primary HDU.')
+            return fits.HDUList([hdu_list[0]])
         else:
-            logger.debug('HDU List does NOT contain a Primary HDU.  Skipping.')
+            logger.debug('HDU List does NOT contain a Primary HDU.'
+                         'Skipping.')
+            return result_hdu_list
 
     def _check_hdu_list(self, cutout_dimensions, hdu_list):
         has_match = False
         len_cutout_dimensions = len(cutout_dimensions)
         if len_cutout_dimensions > 0:
 
+            result_hdu_list = None
             # Check for a pixel cutout
             if isinstance(cutout_dimensions[0], PixelCutoutHDU):
                 if len_cutout_dimensions > 1:
-                    self._write_primary_hdu(hdu_list)
-
+                    result_hdu_list = \
```

Comment:
This isn't really the desired behaviour.  We will get cutouts that have an extension and then the cutout even when we only requested a cutout of a single extension.   hdu_list[0] is always a PrimaryHDU so you will always add the first extension when len_cutout_dimensions > 0, so I'm not sure what you are trying to get at here.

Reply:
Why? `if len_cutout_dimensions > 1:` means that we create the PrimaryHDU only one there are more than 2 cutout extensions.